### PR TITLE
Updates the model to have a seperate log prob for reverse mode

### PR DESCRIPTION
**Pull request by SteveBronder**
#### Submission Checklist

- [x] Run unit tests
- Documentation
    - [ ] If a user-facing facing change was made, the documentation PR is here: <LINK>
    - [x] OR, no user-facing changes were made

## Release notes

This adds a `log_prob_impl` specialization specifically for reverse mode. The main reason for making a separate `log_prob_impl` for reverse mode is to make the code gen simpler when dealing with Stuct of Array style matrices and (hopefully) soon matrices that can go on the GPU.

Instead of having to generate a `conditional_var_value_t<...>` type trait for deciding whether we can use SoA inside of `log_prob`, by making a specialization we can just generate a `stan::math::var_value<...>` whenever we see that a matrix is SoA. We will then be able to do a similar thing for matrices tagged with `GPU`

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by SteveBronder** [comment]
@WardBrian I'll ping you when this is ready for review

File: test/integration/good/code-gen/ode/cpp.expected
Comment (SteveBronder) [inline test/integration/good/code-gen/ode/cpp.expected:286]: This is written in a kind of specific way. I'd like this overload to work if we also use a `stan::math::var_value<Eigen::VectorXd>` or `stan::math::var_value<stan::matrix_cl<double>>`. The latter would be nice if the whole model is on the GPU

**Comment by SteveBronder** [comment]
@WardBrian Alright I think this is ready!

Using an optional type for `reverse_mode_log_prob` made for some kind of goofy logic branches down in the code generation. So instead I just start `reverse_mode_log_prob` with an empty list, then in `transform_mir` I assign the transformed `log_prob` to `reverse_mode_log_prob`. Then we just pass it through the optimizations and code gen.

This always produces a reverse mode specialization for log prob, which imo I think is fine and good. Once we have this in there are several smaller optimizations I'd like to look at doing on the lir that I think would be nice. One example is making the parameters typed `auto` instead of hard types as that will let Eigen matrices and vectors of parameters stay as `Eigen::Map<>` types which never need copied by that math library

**Review by WardBrian — COMMENTED** [review]
A few small questions but this mostly looks good to me.

> Using an optional type for reverse_mode_log_prob made for some kind of goofy logic branches down in the code generation

It's fine if it's not super feasible, but I think there is a (possibly meaningful?) distinction between 'theres nothing here' (`Option.None`) and 'theres something here but its an empty list' so I'd prefer it be an option. By code gen we know it always should be `Some` so that does get an bit annoying

> This always produces a reverse mode specialization for log prob, which imo I think is fine and good

I also think this is fine. It would make it easier to review if it was only for optimizations, but that also makes it more likely for there to be subtle differences we don't notice.

**Comment by WardBrian** [comment]
It's worth grepping around for existing usages of `log_prob` that need to be updated. For example, the `--debug-mem-patterns` flag code:

https://github.com/stan-dev/stanc3/blob/0b1d1cbeb0d4e5cdc1b732bb011340081a3e6a6c/src/analysis_and_optimization/Memory_patterns.ml#L653-L662

**Comment by WardBrian** [comment]
Seems like a copy/paste error led to both overloads now being `require_st_var`

File: src/analysis_and_optimization/Dependence_analysis.ml
Comment (WardBrian) [inline src/analysis_and_optimization/Dependence_analysis.ml:237]: The two new functions here both appear to be dead-code-on-arrival. The version for the normal log prob may also be, since it looks like it is only used in a test.

File: src/middle/Program.ml
Comment (WardBrian) [inline src/middle/Program.ml:134]:
```suggestion
  pp_reverse_mode_log_prob pp_s ppf reverse_mode_log_prob ;
```

File: src/stan_math_backend/Transform_Mir.ml
Comment (WardBrian) [inline src/stan_math_backend/Transform_Mir.ml:722]: I think it's worth adding a comment that right here is the first time reverse_mode_log_prob is initialized, hence assigning it to log prob is not a typo/error

**Review by WardBrian — COMMENTED** [review]
A few small clean ups I missed earlier but then I think this is good to go!

**Review by WardBrian — APPROVED** [review]

**Comment by WardBrian** [comment]
Looks good to me. Do you want to squash out any of the debugging commits before a merge? I'll leave it up to you

**Pull request merged**
